lib/services/iotaRedirector.js

```diff
@@ -26,116 +26,211 @@
 var protocols = require('./protocolData'),
     request = require('request'),
     logger = require('logops'),
-    errors = require('../errors');
+    errors = require('../errors'),
+    _ = require('underscore'),
+    async = require('async');
+
+function guessCollection(body) {
+    if (body.services) {
+        return 'services';
+    } else if (body.devices) {
+        return 'devices';
+    } else {
+        return null;
+    }
+}
+
+function extractProtocols(body) {
+    var collectionName = guessCollection(body);
+
+    function extractProtocol(previous, item) {
+        if (item.protocol) {
+            if (Array.isArray(item.protocol)) {
+                return _.union(previous, item.protocol);
+            } else {
+                previous.push(item);
+
+                return _.uniq(previous);
+            }
+        } else {
+            return previous;
+        }
+    }
+
+    if (collectionName) {
+        return body[collectionName].reduce(extractProtocol, []);
+    } else {
+        return null;
+    }
+}
 
 function queryParamExtractor(req, res, next) {
     if (req.query.protocol) {
-        req.protocolId = req.query.protocol;
+        req.protocolId = [req.query.protocol];
         next();
+    } else if (req.body) {
+        req.splitAndRedirect = true;
+        req.protocolId = extractProtocols(req.body);
+
+        if (req.protocolId) {
+            next();
+        } else {
+            next(new errors.MissingParameters('protocol'));
+        }
     } else {
         next(new errors.MissingParameters('protocol'));
     }
 }
 
-function redirector(req, res, next) {
-    protocols.get(req.protocolId, function(error, protocol) {
-        if (error) {
-            next(error);
+function getProtocols(req, res, next) {
+    function addProtocol(previous, item, callback) {
+        protocols.get(item, function(error, protocol) {
+            if (error) {
+                callback(error);
+            } else {
+                previous[item] = protocol;
+                callback(null, previous);
+            }
+        });
+    }
+
+    async.reduce(req.protocolId, {}, addProtocol, function endReduction(error, result) {
+        req.protocolObj = result;
+        next(error);
+    });
+}
+
+function splitByProtocol(req) {
+    var collectionName = guessCollection(req.body),
+        collections = {};
+
+    for (var i = 0; i < req.body[collectionName].length; i++) {
```

Comment:
It isn't, really, but the reason is somewhat obscure. If you see function `queryParamExtractor()`, you will see that it uses function `extractProtocols()` that in turn uses the previously mentioned function. If the return value of `guessCollection()` is null, so it will be the one for the `extractProtocols()` function, and the original function will raise an error, as you can see in line 75 of the same file.